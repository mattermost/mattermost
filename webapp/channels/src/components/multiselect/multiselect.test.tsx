// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import type {IntlShape} from 'react-intl';

import {renderWithIntl} from 'tests/react_testing_utils';
import {screen, fireEvent} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import MultiSelect from './multiselect';
import type {Value} from './multiselect';
import type {Props as MultiSelectProps} from './multiselect_list';

const element = () => <div/>;

describe('components/multiselect/multiselect', () => {
    const totalCount = 8;
    const optionsNumber = 8;
    const users = [];
    for (let i = 0; i < optionsNumber; i++) {
        users.push({id: `${i}`, label: `${i}`, value: `${i}`});
    }

    const baseProps = {
        ariaLabelRenderer: element as any,
        handleAdd: jest.fn(),
        handleDelete: jest.fn(),
        handleInput: jest.fn(),
        handleSubmit: jest.fn(),
        intl: {} as IntlShape,
        optionRenderer: element,
        options: users,
        perPage: 5,
        saving: false,
        totalCount,
        users,
        valueRenderer: element as any,
        values: [{id: 'id', label: 'label', value: 'value'}],
        valueWithImage: false,
    };

    test('should render multiselect component', () => {
        renderWithIntl(<MultiSelect {...baseProps}/>);
        
        expect(screen.getByRole('combobox')).toBeInTheDocument();
        expect(screen.getByRole('button', {name: /go/i})).toBeInTheDocument();
    });

    test('should handle next page navigation', () => {
        renderWithIntl(<MultiSelect {...baseProps}/>);

        const nextButton = screen.getByRole('button', {name: /next/i});
        fireEvent.click(nextButton);

        expect(screen.getByRole('button', {name: /previous/i})).toBeInTheDocument();
    });

    test('should handle option selection and navigation', () => {
        const renderOption: MultiSelectProps<Value>['optionRenderer'] = (option, isSelected, onAdd, onMouseMove) => {
            return (
                <p
                    key={option.id}
                    data-testid={`option-${option.id}`}
                    onClick={() => onAdd(option)}
                    onMouseMove={() => onMouseMove(option)}
                >
                    {option.id}
                </p>
            );
        };

        const renderValue = (props: {data: {value: unknown}}) => {
            return props.data.value;
        };

        renderWithIntl(
            <MultiSelect
                {...baseProps}
                optionRenderer={renderOption}
                valueRenderer={renderValue}
            />,
        );

        const nextButton = screen.getByRole('button', {name: /next/i});
        fireEvent.click(nextButton);

        // Verify options are rendered
        users.slice(0, baseProps.perPage).forEach((user) => {
            expect(screen.getByTestId(`option-${user.id}`)).toBeInTheDocument();
        });
    });

    test('should render custom no options message', () => {
        const customNoOptionsMessage = (
            <div className='custom-no-options-message'>
                <span>{'No matches found'}</span>
            </div>
        );

        renderWithIntl(
            <MultiSelect
                {...baseProps}
                customNoOptionsMessage={customNoOptionsMessage}
            />,
        );

        expect(screen.getByText('No matches found')).toBeInTheDocument();
    });

    test('should handle back button customization', async () => {
        const handleBackButtonClick = jest.fn();
        
        renderWithIntl(
            <MultiSelect
                {...baseProps}
                backButtonClick={handleBackButtonClick}
                backButtonText='Cancel'
                backButtonClass='tertiary-button'
                saveButtonPosition='bottom'
            />,
        );

        const backButton = screen.getByRole('button', {name: /cancel/i});
        expect(backButton).toHaveClass('tertiary-button');

        await userEvent.click(backButton);
        expect(handleBackButtonClick).toHaveBeenCalled();
    });
});
