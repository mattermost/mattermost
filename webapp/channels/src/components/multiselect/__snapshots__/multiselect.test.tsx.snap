// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/multiselect/multiselect MultiSelectList should match snapshot when custom no option message is defined 1`] = `
<Fragment>
  <div
    className="filtered-user-list"
  >
    <div
      className="filter-row filter-row--full"
    >
      <div
        className="multi-select__container react-select"
      >
        <StateManager
          className=""
          classNamePrefix="react-select-auto react-select"
          components={
            Object {
              "IndicatorsContainer": [Function],
              "Menu": [Function],
              "MultiValueLabel": [Function],
              "MultiValueRemove": [Function],
            }
          }
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          id="selectItems"
          inputValue=""
          isClearable={false}
          isMulti={true}
          menuIsOpen={false}
          onChange={[Function]}
          onInputChange={[Function]}
          onKeyDown={[Function]}
          openMenuOnFocus={false}
          options={
            Array [
              Object {
                "id": "0",
                "label": "0",
                "value": "0",
              },
              Object {
                "id": "1",
                "label": "1",
                "value": "1",
              },
              Object {
                "id": "2",
                "label": "2",
                "value": "2",
              },
              Object {
                "id": "3",
                "label": "3",
                "value": "3",
              },
              Object {
                "id": "4",
                "label": "4",
                "value": "4",
              },
              Object {
                "id": "5",
                "label": "5",
                "value": "5",
              },
              Object {
                "id": "6",
                "label": "6",
                "value": "6",
              },
              Object {
                "id": "7",
                "label": "7",
                "value": "7",
              },
            ]
          }
          styles={
            Object {
              "container": [Function],
            }
          }
          value={
            Array [
              Object {
                "id": "id",
                "label": "label",
                "value": "value",
              },
            ]
          }
        />
        <SaveButton
          disabled={false}
          id="saveItems"
          onClick={[Function]}
          saving={false}
        />
      </div>
      <div
        className="multi-select__help"
        id="multiSelectHelpMemberInfo"
      >
        <MemoizedFormattedMessage
          defaultMessage="{memberOptions, number} of {totalCount, number} members"
          id="multiselect.numMembers"
          values={
            Object {
              "memberOptions": 5,
              "totalCount": 8,
            }
          }
        />
      </div>
    </div>
    <MultiSelectList
      ariaLabelRenderer={[Function]}
      customNoOptionsMessage={
        <div
          className="custom-no-options-message"
        >
          <span>
            No matches found
          </span>
        </div>
      }
      onAction={[Function]}
      onAdd={[Function]}
      onSelect={[Function]}
      optionRenderer={[Function]}
      options={
        Array [
          Object {
            "id": "0",
            "label": "0",
            "value": "0",
          },
          Object {
            "id": "1",
            "label": "1",
            "value": "1",
          },
          Object {
            "id": "2",
            "label": "2",
            "value": "2",
          },
          Object {
            "id": "3",
            "label": "3",
            "value": "3",
          },
          Object {
            "id": "4",
            "label": "4",
            "value": "4",
          },
        ]
      }
      page={0}
      perPage={5}
      query=""
    />
    <div
      className="multi-select__help"
      id="multiSelectMessageNote"
    />
    <div
      className="filter-controls"
    >
      <button
        className="btn btn-sm btn-tertiary filter-control filter-control__next"
        onClick={[Function]}
      >
        <MemoizedFormattedMessage
          defaultMessage="Next"
          id="filtered_user_list.next"
        />
      </button>
    </div>
  </div>
</Fragment>
`;

exports[`components/multiselect/multiselect should match snapshot 1`] = `
<Fragment>
  <div
    className="filtered-user-list"
  >
    <div
      className="filter-row filter-row--full"
    >
      <div
        className="multi-select__container react-select"
      >
        <StateManager
          className=""
          classNamePrefix="react-select-auto react-select"
          components={
            Object {
              "IndicatorsContainer": [Function],
              "Menu": [Function],
              "MultiValueLabel": [Function],
              "MultiValueRemove": [Function],
            }
          }
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          id="selectItems"
          inputValue=""
          isClearable={false}
          isMulti={true}
          menuIsOpen={false}
          onChange={[Function]}
          onInputChange={[Function]}
          onKeyDown={[Function]}
          openMenuOnFocus={false}
          options={
            Array [
              Object {
                "id": "0",
                "label": "0",
                "value": "0",
              },
              Object {
                "id": "1",
                "label": "1",
                "value": "1",
              },
              Object {
                "id": "2",
                "label": "2",
                "value": "2",
              },
              Object {
                "id": "3",
                "label": "3",
                "value": "3",
              },
              Object {
                "id": "4",
                "label": "4",
                "value": "4",
              },
              Object {
                "id": "5",
                "label": "5",
                "value": "5",
              },
              Object {
                "id": "6",
                "label": "6",
                "value": "6",
              },
              Object {
                "id": "7",
                "label": "7",
                "value": "7",
              },
            ]
          }
          styles={
            Object {
              "container": [Function],
            }
          }
          value={
            Array [
              Object {
                "id": "id",
                "label": "label",
                "value": "value",
              },
            ]
          }
        />
        <SaveButton
          disabled={false}
          id="saveItems"
          onClick={[Function]}
          saving={false}
        />
      </div>
      <div
        className="multi-select__help"
        id="multiSelectHelpMemberInfo"
      >
        <MemoizedFormattedMessage
          defaultMessage="{memberOptions, number} of {totalCount, number} members"
          id="multiselect.numMembers"
          values={
            Object {
              "memberOptions": 5,
              "totalCount": 8,
            }
          }
        />
      </div>
    </div>
    <MultiSelectList
      ariaLabelRenderer={[Function]}
      onAction={[Function]}
      onAdd={[Function]}
      onSelect={[Function]}
      optionRenderer={[Function]}
      options={
        Array [
          Object {
            "id": "0",
            "label": "0",
            "value": "0",
          },
          Object {
            "id": "1",
            "label": "1",
            "value": "1",
          },
          Object {
            "id": "2",
            "label": "2",
            "value": "2",
          },
          Object {
            "id": "3",
            "label": "3",
            "value": "3",
          },
          Object {
            "id": "4",
            "label": "4",
            "value": "4",
          },
        ]
      }
      page={0}
      perPage={5}
      query=""
    />
    <div
      className="multi-select__help"
      id="multiSelectMessageNote"
    />
    <div
      className="filter-controls"
    >
      <button
        className="btn btn-sm btn-tertiary filter-control filter-control__next"
        onClick={[Function]}
      >
        <MemoizedFormattedMessage
          defaultMessage="Next"
          id="filtered_user_list.next"
        />
      </button>
    </div>
  </div>
</Fragment>
`;

exports[`components/multiselect/multiselect should match snapshot for page 2 1`] = `
<Fragment>
  <div
    className="filtered-user-list"
  >
    <div
      className="filter-row filter-row--full"
    >
      <div
        className="multi-select__container react-select"
      >
        <StateManager
          className=""
          classNamePrefix="react-select-auto react-select"
          components={
            Object {
              "IndicatorsContainer": [Function],
              "Menu": [Function],
              "MultiValueLabel": [Function],
              "MultiValueRemove": [Function],
            }
          }
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          id="selectItems"
          inputValue=""
          isClearable={false}
          isMulti={true}
          menuIsOpen={false}
          onChange={[Function]}
          onInputChange={[Function]}
          onKeyDown={[Function]}
          openMenuOnFocus={false}
          options={
            Array [
              Object {
                "id": "0",
                "label": "0",
                "value": "0",
              },
              Object {
                "id": "1",
                "label": "1",
                "value": "1",
              },
              Object {
                "id": "2",
                "label": "2",
                "value": "2",
              },
              Object {
                "id": "3",
                "label": "3",
                "value": "3",
              },
              Object {
                "id": "4",
                "label": "4",
                "value": "4",
              },
              Object {
                "id": "5",
                "label": "5",
                "value": "5",
              },
              Object {
                "id": "6",
                "label": "6",
                "value": "6",
              },
              Object {
                "id": "7",
                "label": "7",
                "value": "7",
              },
            ]
          }
          styles={
            Object {
              "container": [Function],
            }
          }
          value={
            Array [
              Object {
                "id": "id",
                "label": "label",
                "value": "value",
              },
            ]
          }
        />
        <SaveButton
          disabled={false}
          id="saveItems"
          onClick={[Function]}
          saving={false}
        />
      </div>
      <div
        className="multi-select__help"
        id="multiSelectHelpMemberInfo"
      >
        <MemoizedFormattedMessage
          defaultMessage="{memberOptions, number} of {totalCount, number} members"
          id="multiselect.numMembers"
          values={
            Object {
              "memberOptions": 3,
              "totalCount": 8,
            }
          }
        />
      </div>
    </div>
    <MultiSelectList
      ariaLabelRenderer={[Function]}
      onAction={[Function]}
      onAdd={[Function]}
      onSelect={[Function]}
      optionRenderer={[Function]}
      options={
        Array [
          Object {
            "id": "5",
            "label": "5",
            "value": "5",
          },
          Object {
            "id": "6",
            "label": "6",
            "value": "6",
          },
          Object {
            "id": "7",
            "label": "7",
            "value": "7",
          },
        ]
      }
      page={1}
      perPage={5}
      query=""
    />
    <div
      className="multi-select__help"
      id="multiSelectMessageNote"
    />
    <div
      className="filter-controls"
    >
      <button
        className="btn btn-sm btn-tertiary filter-control filter-control__prev"
        onClick={[Function]}
      >
        <MemoizedFormattedMessage
          defaultMessage="Previous"
          id="filtered_user_list.prev"
        />
      </button>
    </div>
  </div>
</Fragment>
`;
